### PR TITLE
Common digi packing for CPU and GPU implementations

### PR DESCRIPTION
**Pull request by czangela**
#### PR description:

Follow up to PRs #34509 and #34662.
1. Update documentation of SiPixelDetId
2. Propose merging of CPU and GPU SiPixelDigi packing

If I understood correctly, the GPU implementation doesn't use the packing only to transfer it back to the CPU for validation or other purposes. If so, it makes even more sense to merge the two packing classes.

### CODE REVIEW

**Comment by mmusich** [comment]
allow @czangela test rights

**Comment by czangela** [comment]
@cmsbuild please test

**Comment by czangela** [comment]
enable gpu

**Comment by czangela** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

**Comment by jpata** [comment]
Is this still WIP? If not, please update the title.
Can you also confirm if this was already discussed with the TRK DPG?

**Comment by jpata** [comment]
Pixel (@OzAmram), cc @cms-sw/trk-dpg-l2 let me know if it's OK for you.

**Comment by mmusich** [comment]
@jpata we discussed this with @czangela, it's a preliminary move in order to implement the pixel digi morphing on GPUs.

**Comment by OzAmram** [comment]
Ok for me, would be nice to have the digi morphing on the GPU's too

**Comment by jpata** [comment]
+reconstruction
- technical refactoring of digi packing
- no reco changes expected or observed (GPU also fine)

**Comment by perrotta** [comment]
assign heterogeneous

File: DataFormats/SiPixelDetId/interface/PixelChannelIdentifier.h [hunk 1]
```diff
@@ -2,6 +2,7 @@
 #define DATAFORMATS_PIXELCHANMNELIDENTIFIER_H
 
 #include <utility>
+#include <cstdint>
 
 namespace pixelchanelidentifierimpl {
```
Comment (fwyzard) [inline DataFormats/SiPixelDetId/interface/PixelChannelIdentifier.h:7]: By the way, should this be `pixelchannelidentifierimpl` (with two `n`) ?

File: DataFormats/SiPixelDetId/interface/PixelChannelIdentifier.h [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #define DATAFORMATS_PIXELCHANMNELIDENTIFIER_H
```
Comment (fwyzard) [inline DataFormats/SiPixelDetId/interface/PixelChannelIdentifier.h:2]:
```suggestion
#define  DataFormats_SiPixelDetId_interface_PixelChannelIdentifier_h
```

...or at least fix the typo in `CHANMEL` :-)

**Comment by fwyzard** [comment]
Looking at the code, the changes look fine; I didn't run any tests, but I don't expect any impact on the physics results or timing measurements.

Do you have some Tracking validation plots to double (triple) check that everything is fine ?

**Comment by jpata** [comment]
@czangela kind ping, let us know if the comments by heterogenous can be addressed

File: DataFormats/SiPixelDetId/interface/PixelChannelIdentifier.h (hunk 1 above)
Comment (czangela) [reply DataFormats/SiPixelDetId/interface/PixelChannelIdentifier.h]: I think so :)

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

**Comment by jpata** [comment]
> Do you have some Tracking validation plots to double (triple) check that everything is fine ?

@czangela concerning Andrea's comment above, can you confirm if the GPU tracking validation showed any differences on this PR?

Note that we see some small differences in the GPU DQM test, but I leave to you & other GPU experts to comment if this is expected or not: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_1_X_2021-09-13-1100+19d82a/45399/10824.502_TTbar_13+2018_Patatrack_PixelOnlyGPU+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT/OfflinePixelPV_Alignment.html

**Comment by jpata** [comment]
From having looked at the GPU validation jobs in a few other PRs now, it looks like minor differences in OfflinePixelPV_Alignment are expected, so to reco this looks OK.

I don't see any other changes in the GPU tests that could be relevant for this PR, but please @czangela confirm this directly to heterogenous to address https://github.com/cms-sw/cmssw/pull/34884#issuecomment-913231817.

**Comment by jpata** [comment]
+reconstruction
- for 8dc085a
- stylistic/typo fixes since the signature in https://github.com/cms-sw/cmssw/pull/34884#issuecomment-908983407
- as far as I can tell, GPU validations don't show any relevant changes

**Comment by czangela** [comment]
> From having looked at the GPU validation jobs in a few other PRs now, it looks like minor differences in OfflinePixelPV_Alignment are expected, so to reco this looks OK.
> 
> I don't see any other changes in the GPU tests that could be relevant for this PR, but please @czangela confirm this directly to heterogenous to address [#34884 (comment)](https://github.com/cms-sw/cmssw/pull/34884#issuecomment-913231817).

Yes, I added tracking validation plots here, https://aczirkos.web.cern.ch/aczirkos/validation_34884/ I'll add a readme a bit later.

**Comment by fwyzard** [comment]
Thanks :+1:

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by perrotta** [comment]
+1

**Pull request merged**